File: 0494-target-sum/0494-target-sum.cpp
class Solution {
public:
    int total;
    
    int findTargetSumWays(vector<int>& nums, int S) {
        total = accumulate(nums.begin(), nums.end(), 0);
        
        vector<vector<int>> memo(nums.size(), vector<int>(2 * total + 1, INT_MIN));
        
        return calculate(nums, 0, 0, S, memo);
    }
    
    int calculate(vector<int>& nums, int i, int sum, int S, vector<vector<int>>& memo) {
        if (i == nums.size()) {
            if (sum == S) return 1;
            else return 0;
        }
        
        if (memo[i][sum + total] != INT_MIN) return memo[i][sum + total];
        
        int add = calculate(nums, i + 1, sum + nums[i], S, memo);
        int subtract = calculate(nums, i + 1, sum - nums[i], S, memo);
        return memo[i][sum + total] = add + subtract;   
    }
};

File: 0494-target-sum/0494-target-sum.java
//count partitions such that S1-S2=D(target)


class Solution {
    public int findTargetSumWays(int[] nums, int target) {
        int n=nums.length;
        return countPartitions(n,nums,target);
    }
    public int countPartitions(int n,int[] nums,int target){
        int totSum=0;
        for(int num:nums) totSum+=num;
        int D=target;
        if(totSum-D<0 || (totSum-D)%2==1) return 0;
        int T=(totSum-D)/2;
        int ind=n-1;
        int[][] dp=new int[n][T+1];
        for(int[] row:dp) Arrays.fill(row,-1);
        return ways(ind,nums,T,n,dp);
    }
    public int ways(int ind,int[] nums,int T,int n,int[][] dp){
        if(dp[ind][T]!=-1) return dp[ind][T];
        if(ind==0){
            if(T==0 && nums[0]==0) return 2;
            if(T==0 || nums[0]==T) return 1;
            else return 0;
        }
        int notTake=ways(ind-1,nums,T,n,dp);
        int take=0;
        if(nums[ind]<=T) take=ways(ind-1,nums,T-nums[ind],n,dp);
        return dp[ind][T]=notTake+take;
    }
}